File: articles/calculus-for-systems-modeling/articles/vectors-fields-and-continuous-space/julia/field_audit.jl
scalar_field(x, y) = 20.0 + 2.0 * sin(x) + 0.5 * y * y
vector_field(x, y) = (-y, x)
vector_magnitude(vx, vy) = sqrt(vx*vx + vy*vy)

function audit_field(step, scenario)
    xs = collect(-3.0:step:3.0)
    ys = collect(-3.0:step:3.0)
    scalars = Float64[]
    magnitudes = Float64[]

    for x in xs
        for y in ys
            push!(scalars, scalar_field(x, y))
            vx, vy = vector_field(x, y)
            push!(magnitudes, vector_magnitude(vx, vy))
        end
    end

    warning = step > 0.75 ? "Grid resolution is coarse; field structure may be undersampled." : "Synthetic field audit; document domain, units, and interpolation assumptions."
    return scenario, step, length(scalars), mean(scalars), minimum(scalars), maximum(scalars), mean(magnitudes), maximum(magnitudes), "square domain [-3,3] x [-3,3]", warning
end

println("scenario,grid_step,point_count,scalar_average,scalar_minimum,scalar_maximum,vector_magnitude_average,vector_magnitude_maximum,domain_description,warning")
for case in [(1.0, "coarse_grid"), (0.5, "medium_grid"), (0.25, "fine_grid")]
    println(join(audit_field(case[1], case[2]), ","))
end

File: articles/calculus-for-systems-modeling/articles/vectors-fields-and-continuous-space/r/field_audit.R
scalar_field <- function(x, y) {
  20 + 2 * sin(x) + 0.5 * y^2
}

vector_field <- function(x, y) {
  c(-y, x)
}

vector_magnitude <- function(vx, vy) {
  sqrt(vx^2 + vy^2)
}

audit_field <- function(step, scenario) {
  xs <- seq(-3, 3, by = step)
  ys <- seq(-3, 3, by = step)
  scalars <- c()
  magnitudes <- c()

  for (x in xs) {
    for (y in ys) {
      s <- scalar_field(x, y)
      v <- vector_field(x, y)
      scalars <- c(scalars, s)
      magnitudes <- c(magnitudes, vector_magnitude(v[1], v[2]))
    }
  }

  warning <- ifelse(
    step > 0.75,
    "Grid resolution is coarse; field structure may be undersampled.",
    "Synthetic field audit; document domain, units, and interpolation assumptions."
  )

  data.frame(
    scenario = scenario,
    grid_step = step,
    point_count = length(scalars),
    scalar_average = mean(scalars),
    scalar_minimum = min(scalars),
    scalar_maximum = max(scalars),
    vector_magnitude_average = mean(magnitudes),
    vector_magnitude_maximum = max(magnitudes),
    domain_description = "square domain [-3,3] x [-3,3]",
    warning = warning
  )
}

results <- rbind(
  audit_field(1.0, "coarse_grid"),
  audit_field(0.5, "medium_grid"),
  audit_field(0.25, "fine_grid")
)

dir.create("outputs/tables", recursive = TRUE, showWarnings = FALSE)
write.csv(results, "outputs/tables/r_field_audit.csv", row.names = FALSE)
print(results)
